### DISCUSSION TITLE
v2.3.0-A – Community Builds Thread

### DISCUSSION THREAD

**Discussion by gianni-rosato** [gianni-rosato/svt-av1-psy#118] [Show and tell]
# Community Builds Thread

This is a place for the community to share unofficial tools not affiliated with the project — mainly consisting of binaries compiled by community members.

## Trust

:warning: **Community-built binaries are not affiliated with the SVT-AV1-PSY project & are to be used at your own risk.** If you encounter bugs while using a community-built binary, ensure that you double-check with the person responsible for providing you with the binary before submitting an issue report.

## Architecture

When downloading pre-compiled binaries, you might see AVX, AVX2, AVX-512, x86-64-v3, etc. If you don't know exactly what ISA extensions your CPU supports, here is a chart to help you quickly understand your hardware's support:

### Not Vendor-Specific
```c
x86-64    : Processors lacking SSE4.2 and newer features
x86-64-v2 : Processors with features up to and including SSE4.2
x86-64-v3 : Processors with features up to and including AVX and/or AVX2
x86-64-v4 : Processors with features up to and including AVX512
```

ISA extensions like AVX, AVX2, & AVX512 are for *vector processing* which SVT-AV1-PSY relies heavily on for its fast multithreaded performance. Below are some helpful charts to help you narrow down the best options if you have popular x64 hardware:

### AMD
```ocaml
Generation                                             Vectors  Build
Zen   (1xxx)                                           AVX2     x86-64-v3 / znver1
Zen+  (1xxxAF-2xxx-3xxxG/H/U/C)                        AVX2     x86-64-v3 / znver1
Zen2  (3xxx-4xxxS/G/H/U-5700U-5500U-5300U-7520U-7320U) AVX2     x86-64-v3 / znver2
Zen3  (5xxx-7330U-7530U-7730U)                         AVX2     x86-64-v3 / znver3
Zen3+ (6xxx-7xxxH/U)                                   AVX2     x86-64-v3 / znver3
Zen4  (7xxx-8xxxG)                                     AVX512   x86-64-v4 / znver4
```
### Intel (Desktop)
```js
Generation                Vectors  Build
Haswell     (4th gen)     AVX2     x86-64-v3 / haswell
Broadwell   (5th gen)     AVX2     x86-64-v3 / broadwell
Skylake     (6th gen)     AVX2     x86-64-v3 / skylake
Kaby Lake   (7th gen)     AVX2     x86-64-v3 / skylake
Coffee Lake (8-9th gen)   AVX2     x86-64-v3 / skylake
Comet Lake  (10th gen)    AVX2     x86-64-v3 / skylake
Rocket Lake (11th gen)    AVX512   x86-64-v4 / rocketlake
Alder Lake  (12th gen)    AVX2     x86-64-v3 / alderlake
Raptor Lake (13-14th gen) AVX2     x86-64-v3 / raptorlake
```

Included below is even more information about what is available when specifying `-march` & `-mtune` on x64 CPUs:

Known valid x64 arguments for `-march=`:
> i386 i486 i586 pentium lakemont pentium-mmx winchip-c6 winchip2 c3 samuel-2 c3-2 nehemiah c7 esther i686 pentiumpro pentium2 pentium3 pentium3m pentium-m pentium4 pentium4m prescott nocona core2 nehalem corei7 westmere sandybridge corei7-avx ivybridge core-avx-i haswell core-avx2 broadwell skylake skylake-avx512 cannonlake icelake-client rocketlake icelake-server cascadelake tigerlake cooperlake sapphirerapids emeraldrapids alderlake raptorlake meteorlake graniterapids graniterapids-d arrowlake arrowlake-s lunarlake pantherlake bonnell atom silvermont slm goldmont goldmont-plus tremont gracemont sierraforest grandridge clearwaterforest knl knm intel geode k6 k6-2 k6-3 athlon athlon-tbird athlon-4 athlon-xp athlon-mp x86-64 x86-64-v2 x86-64-v3 x86-64-v4 eden-x2 nano nano-1000 nano-2000 nano-3000 nano-x2 eden-x4 nano-x4 lujiazui yongfeng k8 k8-sse3 opteron opteron-sse3 athlon64 athlon64-sse3 athlon-fx amdfam10 barcelona bdver1 bdver2 bdver3 bdver4 znver1 znver2 znver3 znver4 znver5 btver1 btver2 generic native

Known valid x64 arguments for `-mtune=`:
> generic i386 i486 pentium lakemont pentiumpro pentium4 nocona core2 nehalem sandybridge haswell bonnell silvermont goldmont goldmont-plus tremont sierraforest grandridge clearwaterforest knl knm skylake skylake-avx512 cannonlake icelake-client icelake-server cascadelake tigerlake cooperlake sapphirerapids alderlake rocketlake graniterapids graniterapids-d arrowlake arrowlake-s pantherlake intel lujiazui yongfeng geode k6 athlon k8 amdfam10 bdver1 bdver2 bdver3 bdver4 btver1 btver2 znver1 znver2 znver3 znver4 znver5

- Note: `-march=foo` implies `-mtune=foo` unless you also specify a different `-mtune`. This is one reason why using `-march` is better than just enabling options like `-mavx` without doing anything about tuning.

## Antivirus

🛑 Be wary of antivirus software on Windows detecting EXEs distributed here as malicious software. While they may not always legitimately be malicious, it is important to maintain a healthy level of skepticism when running code that someone else has compiled.

**Comment by ItachiUchiha-IU** [comment]
Hi, coincidentally I happen to not be at home again during a new release, so I'm gonna post the builds in remote from my phone again.

This time there's some informations that need to be given, because this new version ends up in non-stable binaries for windows:
1) v2.3.0-A is slower than the previous version, especially if not using `--psy-rd` (or using very low values); as confirmed with the "PSY guys" (the devs that work on this fork), this is the expected behavior as the `--psy-rd` code runs and it's called quite often; thus, with it off or with such low values, from tests made on different machines on both linux and windows, it can be slower by 10-20%
- "solution" : use `--psy-rd 0.1` or higher, and the speed difference will go down a lot (in some of my tests on windows I got down to 2-3% speed diff, while on linux it seems to be working a little better in some cases); important info about `--psy-rd` > 0 at point 3

2) the binaries for v2.3.0-A built normally on windows crash; injecting the binary to redirect the calls to an external mimalloc seems to be fixing windows builds (some users are reporting crashes with their injected builds on 8-bit encodes only, but idk if this affects mine too, as 8-bit av1 is **illegal** anyway), thus, I have now injected all my builds; u'll need to keep the mimalloc files in the same folder as the svtav1encapp.exe binary; the cross-compiled binary doesn't seem to be crashing for 8-bit encoding either but it's quite slower than the windows builds: some got an 8% speed diff, some way more (I can't test myself if it crashes or what speed diff I get cause of the reasons mentioned above)
- ~~"solution" : use av1an. av1an splits the encode per scene before encoding them, and if a chunk crashes, it simply re-encodes that single chunk and the encoding continues without any major problem~~ this point doesn't stand in the cases, with specific settings, where the encoder crashes instantly cause that means it keeps crashing before even completing a single chunk and thus av1an can't save da day; so I'll just cross out this point
- av1an is always recommended as it is able to keep encoding even if the encoder crashes occasionally (and for many other reasons like full cpu usage, zoning, better scene detection and keyframe choices, stopping and resuming the encoding)
- the psy guys already know about the binaries crashing on windows cause the only ones reporting these crashes are the ones of us that are on windows. we have already started looking into it together with them but it could still end up being something not easy to fix, so it may be something that will get worked on tog with mainline after v3 is released.

3) `--psy-rd` ON (which means any value > 0) is currently broken on windows and it will [create artifacts](https://github.com/psy-ex/svt-av1-psy/issues/117) (it can actually have [artifacts on linux too](https://github.com/psy-ex/svt-av1-psy/issues/117#issuecomment-2567603357))
- "solution" : turn `--psy-rd` OFF (0) or use p-1 as it seems to be fixing the artifacts (actually idk if p0 or p1 fix it too, as I only tested p-1 and p2+, or if higher crf values than the 18 I tested create artifacts again, so be careful about it; u can use video-compare or vspreview to compare two videos easily and search for eventual artifacts)

4) I don't remember if I said everything I wanted to say, so if something else comes to mind later, or something else comes out, I'll edit it and/or add points.

More discussions regarding this are in the v2.3.0 discussion -> https://github.com/gianni-rosato/svt-av1-psy/discussions/88#
Infos about the builds are at https://github.com/gianni-rosato/svt-av1-psy/discussions/88# and https://github.com/gianni-rosato/svt-av1-psy/discussions/69 for now.
~~[svt-av1-psy_2024-12-25_20.29.49_MULTI-MARCH_clang19.1.5_2.zip](https://github.com/user-attachments/files/18259726/svt-av1-psy_2024-12-25_20.29.49_MULTI-MARCH_clang19.1.5_2.zip)
[svt-av1-psy_2024-12-25_20.42.02_MULTI-MARCH_av1an-pgo_clang19.1.5.zip](https://github.com/user-attachments/files/18259728/svt-av1-psy_2024-12-25_20.42.02_MULTI-MARCH_av1an-pgo_clang19.1.5.zip)
[svt-av1-psy_2024-12-25_21.08.08_MULTI-MARCH_dovi+hdr10plus_clang19.1.5_2.zip](https://github.com/user-attachments/files/18259730/svt-av1-psy_2024-12-25_21.08.08_MULTI-MARCH_dovi%2Bhdr10plus_clang19.1.5_2.zip)
[svt-av1-psy_2024-12-25_23.13.06_MULTI-MARCH_dovi+hdr10plus_av1an-pgo_clang19.1.5_2.zip](https://github.com/user-attachments/files/18259733/svt-av1-psy_2024-12-25_23.13.06_MULTI-MARCH_dovi%2Bhdr10plus_av1an-pgo_clang19.1.5_2.zip)~~
Injecting the binary to redirect the calls to mimalloc (`.\minject.exe -i .\SvtAv1EncApp.exe`) seems to fix the problem on windows builds (credits to @Andarwinux), so I invite u to try the following injected builds (u need the mimalloc dlls in the same folder as svtav1encapp.exe) (encoding in 8-bit may still result in crashes but honestly if u're encoding in av1 8-bit u're doing something wrong at the base, as av1 is only developed at 10-bit):
[svt-av1-psy_2024-12-25_20.29.49_MULTI-MARCH_clang19.1.5_injected_2.zip](https://github.com/user-attachments/files/18266627/svt-av1-psy_2024-12-25_20.29.49_MULTI-MARCH_clang19.1.5_injected_2.zip)
[svt-av1-psy_2024-12-27_12.20.00_MULTI-MARCH_av1an-pgo_clang19.1.5_injected_2.zip](https://github.com/user-attachments/files/18266593/svt-av1-psy_2024-12-27_12.20.00_MULTI-MARCH_av1an-pgo_clang19.1.5_injected_2.zip)
[svt-av1-psy_2024-12-25_21.08.08_MULTI-MARCH_dovi+hdr10plus_clang19.1.5_injected_2.zip](https://github.com/user-attachments/files/18266838/svt-av1-psy_2024-12-25_21.08.08_MULTI-MARCH_dovi%2Bhdr10plus_clang19.1.5_injected_2.zip)
[svt-av1-psy_2024-12-25_23.13.06_MULTI-MARCH_dovi+hdr10plus_av1an-pgo_clang19.1.5_injected_2.zip](https://github.com/user-attachments/files/18266842/svt-av1-psy_2024-12-25_23.13.06_MULTI-MARCH_dovi%2Bhdr10plus_av1an-pgo_clang19.1.5_injected_2.zip)
Any report is appreciated.

**Reply by igowrek** [reply]
I'm getting no measurable slowdown when using `--psy-rd 6.0` vs `--psy-rd 0`, only an overall 2-3% slowdown vs 2.3.0, at least on my builds.
As far as crashing goes, mine crashes almost instantly when trying to encode in 8-bit.
10-bit crashes much later, tens of minutes into the encode, and not always.

**Reply by ItachiUchiha-IU** [reply]
Thx for the report!
I am starting to think the slowdown with psy-rd off and low values may be happening together with some other setting.

Btw I added a binary that may not be crashing, if anyone is able to test it and give a report, thx!

**Reply by Andarwinux** [reply]
For me all builds crash when with --preset<=6

**Reply by Andarwinux** [reply]
I've found an extremely weird phenomenon: if I use [minject](https://github.com/microsoft/mimalloc/tree/dev/bin) to redirect the malloc calls of these builds to mimalloc they don't crash anyway, otherwise even my mingw builds crash.

**Reply by ItachiUchiha-IU** [reply]
Thx for the reports, and yeah I got reported it works partially for someone and crashes instantly on someone else, sometimes depending on the clips sometimes on the settings.

I guess I'll try some absurd training just for the lulz and see if it helps.
Sadly on remote I can only try to some extent.

**Reply by ItachiUchiha-IU** [reply]
> As far as crashing goes, mine crashes almost instantly when trying to encode in 8-bit.

Just as a reminder, 8-bit is "banned" from av1. As it's never been developed and performs hugely worse than 10-bit and should never be used.

**Reply by Andarwinux** [reply]
commit 129faba15d1d628658ee84f7e4ebce28c716d493 caused this crash

**Reply by Andarwinux** [reply]
```
==10276==ERROR: AddressSanitizer: access-violation on unknown address 0x010078dcb000 (pc 0x7ff6be039167 bp 0x0100a7dfc6e0 sp 0x0100a7dfc640 T61)
==10276==The signal is caused by a READ memory access.
    #0 0x7ff6be039166 in SpatialFullDistortionKernel64_AVX512_INTRIN /svt-av1-psy/Source/Lib/ASM_AVX512/pic_operators_intrin_avx512.c:158
    #1 0x7ff6be037745 in svt_spatial_full_distortion_kernel_avx512 /svt-av1-psy/Source/Lib/ASM_AVX512/pic_operators_intrin_avx512.c:314
    #2 0x7ff6bf2ffa1a in picture_sse_calculations /svt-av1-psy/Source/Lib/Codec/deblocking_filter.c:770
    #3 0x7ff6bf2feaaa in try_filter_frame /svt-av1-psy/Source/Lib/Codec/deblocking_filter.c:935
    #4 0x7ff6bf2fd775 in search_filter_level /svt-av1-psy/Source/Lib/Codec/deblocking_filter.c:989
    #5 0x7ff6bf2fc9d9 in svt_av1_pick_filter_level /svt-av1-psy/Source/Lib/Codec/deblocking_filter.c:1299
    #6 0x7ff6bdaecfd1 in svt_aom_dlf_kernel /svt-av1-psy/Source/Lib/Codec/dlf_process.c:104
    #7 0x7ff8436f3d6d in _asan_wrap_CreateThread+0x14d (D:\sw\bin\clang_rt.asan_dynamic.dll+0x180053d6d)
    #8 0x7ff8a9d655a6 in BaseThreadInitThunk+0x16 (C:\WINDOWS\System32\KERNEL32.DLL+0x1800355a6)
    #9 0x7ff8ab37491f in RtlUserThreadStart+0x1f (C:\WINDOWS\SYSTEM32\ntdll.dll+0x1800f491f)

==10276==Register values:
rax = 10078dcafd4  rbx = 100a7dfd200  rcx = 1ba55d00000  rdx = 10078dcb013
rdi = 0  rsi = 420  rbp = 100a7dfc6e0  rsp = 100a7dfc640
r8  = 100a7dfd080  r9  = f2f2f2f2f2f2f2f2  r10 = 438  r11 = 780
r12 = 0  r13 = 0  r14 = 0  r15 = 0
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: access-violation /svt-av1-psy/Source/Lib/ASM_AVX512/pic_operators_intrin_avx512.c:158 in SpatialFullDistortionKernel64_AVX512_INTRIN
Thread T61 created by T0 here:
    #0 0x7ff8436f3c82 in _asan_wrap_CreateThread+0x62 (D:\sw\bin\clang_rt.asan_dynamic.dll+0x180053c82)
    #1 0x7ff6bdb12191 in svt_create_thread /svt-av1-psy/Source/Lib/Codec/svt_threads.c:125
    #2 0x7ff6bd80f76d in svt_av1_enc_init /svt-av1-psy/Source/Lib/Globals/enc_handle.c:2699
    #3 0x7ff6bd7d0684 in init_encoder /svt-av1-psy/Source/App/app_context.c:424
    #4 0x7ff6bd7d74e4 in enc_context_ctor /svt-av1-psy/Source/App/app_main.c:238
    #5 0x7ff6bd7d632a in main /svt-av1-psy/Source/App/app_main.c:519
    #6 0x7ff6bf3a3918 in invoke_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
    #7 0x7ff6bf3a3a51 in __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    #8 0x7ff6bf3a3add in __scrt_common_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:330
    #9 0x7ff6bf3a3afd in mainCRTStartup D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp:16
    #10 0x7ff8a9d655a6 in BaseThreadInitThunk+0x16 (C:\WINDOWS\System32\KERNEL32.DLL+0x1800355a6)
    #11 0x7ff8ab37491f in RtlUserThreadStart+0x1f (C:\WINDOWS\SYSTEM32\ntdll.dll+0x1800f491f)

==10276==ABORTING
```

**Reply by igowrek** [reply]
> I've found an extremely weird phenomenon: if I use [minject](https://github.com/microsoft/mimalloc/tree/dev/bin) to redirect the malloc calls of these builds to mimalloc they don't crash anyway, otherwise even my mingw builds crash.

Can confirm no crash on 8-bit encode with minjected mimalloc on native clang-cl build

**Reply by Ironclad17** [reply]
These were built with ninja and clang-cl: `/clang:-O2 -flto`. IR+CSIR PGO
[2.3.0-A.svt-av1-psy.dovi.hdr10plus.mimalloc.znver2.clang19.1.6.zip](https://github.com/user-attachments/files/18268706/2.3.0-A.svt-av1-psy.dovi.hdr10plus.mimalloc.znver2.clang19.1.6.zip)
[2.3.0-A.svt-av1-psy.dovi.hdr10plus.mimalloc.pgo.znver2.clang19.1.6.zip](https://github.com/user-attachments/files/18268707/2.3.0-A.svt-av1-psy.dovi.hdr10plus.mimalloc.pgo.znver2.clang19.1.6.zip)
Update: Mimalloc resolved all issues for me. 8bit and 10bit clips encoded fine in both versions. All dlls included.

**Reply by ItachiUchiha-IU** [reply]
I just had time to inject my builds and they seem to be working now (tested znver4 pgo and non-pgo).
If anyone can test them and report, I would appreciate.
I also thank everyone who participated in trying to find a solution to the windows crashes.
I have just read that ppl are still experiencing crashes with their injected builds with 8-bit content(or encoding in 8-bit? why would anyone encode in 8-bit with av1 tho).

**Reply by tuanden0** [reply]
@ItachiUchiha-IU 

I used your "svt-av1-psy_2024-12-25_20.29.49_MULTI-MARCH_clang19.1.5_injected_2.zip", and try to encode with 4 instances and it printed out "out of memory" on the console but used only 15% on my machine. I rollback to v2.3.0 and can run 4 instances without any problem.

![image](https://github.com/user-attachments/assets/d45bf17a-4abe-484b-a5c0-578ea6348ec2)

**Reply by ItachiUchiha-IU** [reply]
>  tried to encode with 4 instances and it printed out "out of memory" on the console but used only 15% on my machine. I rollback to v2.3.0 and can run 4 instances without any problem.

You mean u're contemporary using 4 instances of the encoder to saturate your CPU because with one instance it works correctly but has very low CPU usage?
If I understood it correctly, I'd strongly recommend to use [Av1an](https://github.com/master-of-zen/Av1an) instead.

If u need guidance to install and use av1an, don't worry and ask here.

**Reply by tuanden0** [reply]
@ItachiUchiha-IU 

Yes, I run 4 instances of the encoder to saturate my CPU. 

But on my report, the error is printed out it "out of memory" and exit right away but my RAM usage is about 15%.

Thank for your recommend, I will try to use Av1an.

**Reply by ItachiUchiha-IU** [reply]
@tuanden0 we just found out that mimalloc did largerly increase RAM usage on injected builds
Idk what build u're using rn and if u still need it, but this should fix the problem with the absurd RAM usage (and thus allow u to use builds injected with mimalloc): https://github.com/orgs/psy-ex/discussions/128#discussioncomment-12296702

**Reply by Andarwinux** [reply]
Try to increase the swapfile, mimalloc prefers to use eager commit to maximize performance, which consumes a lot of virtual address space, and since Windows doesn't implement overcommit, it requires enough SWAP.

**Reply by ItachiUchiha-IU** [reply]
I set purge delay to 0

**Comment by Uranite** [comment]
Termux build for arm and aarch64 with flto
Pick one according to your arch, and install via `dpkg -i`
Afterwards, you can encode via libsvtav1 in ffmpeg
Disclaimer: the build is actually https://github.com/psy-ex/svt-av1-psy/commit/f560b40c1de42f1ca3ea35f575be4cfc24083dcc instead of v2.3.0-A, but I don't feel like rebuilding it
[termux-svt-av1-psy-v2.3.0-A.zip](https://github.com/user-attachments/files/18293340/termux-svt-av1-psy-v2.3.0-A.zip)

**Comment by igowrek** [comment]
I found that znver3 is not slower than znver2 when building with IR+CSIR PGO, binary has a comparable footprint as well.
Also, for the first time I'm seeing a slight improvement instead of slowdown when using `-O3` over `-O2`.
Will be sharing multiple builds for various architectures that use profiling data generated for O3 znver3.

Edit: here are all the builds, many marches, including generic.
I used the current state testing branch because I was lazy to switch back to master when I was generating profiling data 😁
Also includes ~~statically~~ dynamically linked **mimalloc** and optional Dolby Vision + HDR10+ support.
~~[2.3.0-A(testing).IR+CSIR.PGO.-O3.all.marches.mimalloc.[dovi+hdr10+].zip](https://github.com/user-attachments/files/18301101/2.3.0-A.testing.IR%2BCSIR.PGO.-O3.all.marches.mimalloc.dovi%2Bhdr10%2B.zip)~~
[2.3.0-A(testing).IR+CSIR.PGO.-O3.all.marches.dyn.mimalloc.[dovi+hdr10+].Clang.19.1.6.zip](https://github.com/user-attachments/files/18304617/2.3.0-A.testing.IR%2BCSIR.PGO.-O3.all.marches.dyn.mimalloc.dovi%2Bhdr10%2B.Clang.19.1.6.zip)

Dynamically linked mimalloc lib gave me a **massive** boost of **~2.5%❗** over no mimalloc/failed statically linked.

**Reply by Uranite** [reply]
I'll be sure to test those builds once you share them

**Reply by Andarwinux** [reply]
It would be nice if you could disable SSE4A, otherwise I can't test it on any Intel machine.

**Reply by igowrek** [reply]
> It would be nice if you could disable SSE4A, otherwise I can't test it on any Intel machine.

I'm gonna include lots of Intel marches, so I think it should be ok, right?

**Reply by igowrek** [reply]
@Uranite @Andarwinux the builds are here!

**Reply by Andarwinux** [reply]
Statically linked mimalloc doesn't seem to work, MIMALLOC_VERBOSE gives empty statistics.

Your build doesn't seem to have a significant performance advantage on my machine (tigerlake). It looks like CSIR is not suitable for cross-microarchitecture? As far as I know CSIR optimization and instrumentation happens after inlining and before SelectionDAG, so it is strongly correlated with machine instructions rather than IR, which I guess leads to a lot of hot-cold function mispredictions.

**Reply by igowrek** [reply]
> Statically linked mimalloc doesn't seem to work, MIMALLOC_VERBOSE gives empty statistics.

It is indeed not working. I just checked vs injected. I'll try to fix that.

**Reply by igowrek** [reply]
@Andarwinux looks like mimalloc just doesn't do much when doing a simple static link, I got it to show some stats, but it's still not doing anything really, so I just linked dynamically instead. I think it wasn't really meant to work on Windows that way.

**Reply by igowrek** [reply]
Pitted my znver3 -O3 mimalloc build against @ItachiUchiha-IU's znver2 -O2 mimalloc build.
Mine got a 5.5% speed advantage.
Now I gotta test some completely random clip.
Edit: the random unrelated clips and settings aren't doing as great, ~2.5-3.5% increase.

**Reply by ItachiUchiha-IU** [reply]
I hope u're compiling the binaries with the same params apart from the ones u mentioned (no `-ffastmath` etc) when comparing speeds, and using av1an or a method that assures real 100% cpu utilization

**Reply by igowrek** [reply]
> I hope u're compiling the binaries with the same params apart from the ones u mentioned (no `-ffastmath` etc) when comparing speeds, and using av1an or a method that assures real 100% cpu utilization

From my testing -ffast-math doesn't really do anything, maybe makes a binary just a tiny bit smaller. I throw it in just because, but it's not really helping with performance. I just compared on vs off, and there was no difference.
I also used chunked encoding to compare performance.

**Reply by igowrek** [reply]
@ItachiUchiha-IU actually, I just noticed an increase in performance from `--ffast-math` of ~2% when using more than 2 chunks 🤔
I hope StaxRip implements chunked RPU's soon, so I can actually take advantage of that, cause right now when encoding with DoVi you'd have to encode the whole file at once in order to properly inject the RPU. That is unless you're willing to manually split the RPU and then separately schedule all those chunks for encoding with respective source chunks.

**Reply by igowrek** [reply]
I also found that it doesn't hurt to generate more profiles than just one, at least when it comes to presets.
I usually generate for preset 4, but now I tried to generate for the 2-6 range of presets.
Preset 4 performs exactly the same with many profiles as with preset 4 only profile.

And at the same time, doing just preset 2, like the build.sh PGO build does, results in worse performance for preset 4 than if I were to also generate profile data for preset 4.

Also, PGO in build.sh uses `--sparse` profdata merging, which supposedly provides better performance and increases binary size, but in the end only the latter is true and performance is the same.

I think I'm gonna find how far I can get this with profiles, to generate as much as possible. I already found out that source resolution doesn't matter for profile generation, I went down from 1920x800 to 240x100 and results are indistinguishable, so that saves a bunch of time at least.
Upd: Actually, no, looks like resolution can be too low, it hurts performance by a tiny bit at 240x100, but not at 480x200, gonna test 360x150.
Next I'll be trying to add the whole range of presets, not just 2-6. And I'm already doing all combinations of presets 2-6 with tunes 0-3.
360x150 was also a bit slower. Imma retest 480x200 and get on with it.
Upd:Upped resolution a bit to 720x300, didn't like the retest run on 480x200.
Doing profiling with presets -1 through 11 now, dropped tune 0, cause I don't care much about that one (should I?)

**Reply by igowrek** [reply]
@ItachiUchiha-IU I found chunked encoding to be not a very reliable benchmark.
When chunks are being encoded, their CPU usage is not cosistent run to run, resulting in one of the chunks finishing faster in one run, but then slow in another run. This can result in one chunk being encoded slow enough in some runs to skew the benchmark result by a few seconds.
For it to be more or less reliable, it would have to be run for much longer, so that the resource allocation to each chunk is evened out over time.

On the other hand, whole file encoding is always using all resources it's able to use.
Coupled with limiting boost frequency to the point where thermal throttling is irrelevant, it's a much more precise benchmark.

**Reply by igowrek** [reply]
I looked into Av1an, I like the concept, really fair benchmark in terms of load with enough workers, since each job is rather short and gets shorter to the end of encode process.
But the scene detection process is rather long. And what I found, was that at preset 4 for a 1920x800 movie downscaled from 3840x1600, the time it takes to encode the whole thing with Av1an is about the same as it took with a single SvtAv1EncApp job in StaxRip, when I limit the power to 90W PPT on my Ryzen 5 5700X for both cases.
And another thing is that DoVi can't really be used with Av1an, unless someone implements a way to split the RPU based on chunks defined by Av1an and feed that to it.

**Reply by Andarwinux** [reply]
> Also, PGO in build.sh uses `--sparse` profdata merging, which supposedly provides better performance and increases binary size, but in the end only the latter is true and performance is the same.

This removes empty counters from profdata to ensure that the performance of cold paths is not affected by the profile guided size optimization, which is not needed if your training process is sufficiently complete and representative. However, I think this will prevent serious regressions in preset 1.

**Reply by Uranite** [reply]
Sorry for the late reply, I was busy and when I had the time, I decided to do something else. Anyway, on my pc, your build is 1.3% slower than Itachi Uchiha's build, or 2.5% slower than my own PGO build, though I noticed that your build is actually https://github.com/psy-ex/svt-av1-psy/commit/fbe4e20af7b7b3bfa0a957e217feb2409f849fdc and not v2.3.0-A, so I didn't actually test the same version.

**Reply by igowrek** [reply]
> Sorry for the late reply, I was busy and when I had the time, I decided to do something else. Anyway, on my pc, your build is 1.3% slower than Itachi Uchiha's build, or 2.5% slower than my own PGO build, though I noticed that your build is actually [fbe4e20](https://github.com/psy-ex/svt-av1-psy/commit/fbe4e20af7b7b3bfa0a957e217feb2409f849fdc) and not v2.3.0-A, so I didn't actually test the same version.

Yeah, I later tested against v2.3.0-A and the testing is a bit slower than v2.3.0-A
